src/IRMatch.h

```diff
@@ -45,6 +47,1366 @@ bool expr_match(Expr pattern, Expr expr, std::map<std::string, Expr> &result);
+    s << i;
+    return s;
+}
+
+template<typename Op>
+int64_t constant_fold_bin_op(halide_type_t, int64_t, int64_t, uint32_t &);
+
+template<typename Op>
+uint64_t constant_fold_bin_op(halide_type_t, uint64_t, uint64_t, uint32_t &);
+
+template<typename Op>
+double constant_fold_bin_op(halide_type_t, double, double, uint32_t &);
+
+template<typename Op>
+uint64_t constant_fold_cmp_op(halide_type_t, int64_t, int64_t);
+
+template<typename Op>
+uint64_t constant_fold_cmp_op(halide_type_t, uint64_t, uint64_t);
+
+template<typename Op>
+uint64_t constant_fold_cmp_op(halide_type_t, double, double);
+
+
+// Matches one of the binary operators
+template<typename Op, typename A, typename B>
+struct BinOp {
+    struct pattern_tag {};
+    A a;
+    B b;
+    HALIDE_ALWAYS_INLINE bool match(const BaseExprNode &e, MatcherState &state) const {
+        if (e.node_type != Op::_node_type) {
+            return false;
+        }
+        const Op &op = (const Op &)e;
+        return a.match(*op.a.get(), state) && b.match(*op.b.get(), state);
+    }
+
+    template<typename Op2, typename A2, typename B2>
+    HALIDE_ALWAYS_INLINE bool match(const BinOp<Op2, A2, B2> &op, MatcherState &state) const {
+        return std::is_same<Op, Op2>::value && a.match(op.a, state) && b.match(op.b, state);
+    }
+
+    HALIDE_ALWAYS_INLINE Expr make(MatcherState &state) const {
+        Expr ea = to_expr(a, state), eb = to_expr(b, state);
+        match_types(ea, eb);
+        return Op::make(std::move(ea), std::move(eb));
+    }
+
+    template<typename A1 = A,
+             typename B1 = B,
+             typename = decltype(std::declval<A1>().make_folded_const(std::declval<MatcherState &>())),
+             typename = decltype(std::declval<B1>().make_folded_const(std::declval<MatcherState &>()))>
+    HALIDE_ALWAYS_INLINE MatcherState::Const make_folded_const(MatcherState &state) const {
+        MatcherState::Const ca = a.make_folded_const(state);
+        MatcherState::Const cb = b.make_folded_const(state);
+        MatcherState::Const result;
+        result.type = ca.type;
+        result.flags = ca.flags | cb.flags;
+        switch (ca.type.code) {
+        case halide_type_int:
+            result.val.i = constant_fold_bin_op<Op>(result.type, ca.val.i, cb.val.i, result.flags);
+            break;
+        case halide_type_uint:
+            result.val.u = constant_fold_bin_op<Op>(result.type, ca.val.u, cb.val.u, result.flags);
+            break;
+        case halide_type_float:
+            result.val.f = constant_fold_bin_op<Op>(result.type, ca.val.f, cb.val.f, result.flags);
+            break;
+        default:
+            // unreachable
+            ;
+        }
+        return result;
+    }
+};
+
+// Matches one of the comparison operators
+template<typename Op, typename A, typename B>
+struct CmpOp {
+    struct pattern_tag {};
+    A a;
+    B b;
+    HALIDE_ALWAYS_INLINE bool match(const BaseExprNode &e, MatcherState &state) const {
```

Comment:
For all of the state-by-ref calls here, it would be nice to have a comment about whether it's in+outparam, or purely outparam (I find myself forgetting as I read the code)